File: solutions/go/ocr-numbers/1/ocr_numbers.go
package ocr

import "strings"

var digit = map[string]byte{
	" _ | ||_|   ": '0',
	"     |  |   ": '1',
	" _  _||_    ": '2',
	" _  _| _|   ": '3',
	"   |_|  |   ": '4',
	" _ |_  _|   ": '5',
	" _ |_ |_|   ": '6',
	" _   |  |   ": '7',
	" _ |_||_|   ": '8',
	" _ |_| _|   ": '9',
}

func recognizeDigit(lines []string, start int) byte {
	if len(lines) < 4 {
		return '?'
	}
	need := start + 3
	for _, line := range lines {
		if len(line) < need {
			return '?'
		}
	}
	if d, ok := digit[lines[0][start:need]+
		lines[1][start:need]+
		lines[2][start:need]+
		lines[3][start:need]]; ok {
		return d
	}
	return '?'
}

func Recognize(s string) []string {
	lines := strings.Split(s, "\n")
	if lines[0] == "" {
		lines = lines[1:]
	}
	r := make([]string, (len(lines)+3)/4)
	for index := range r {
		max_ := 0
		for lineIndex := 0; lineIndex < 4 && lineIndex < len(lines); lineIndex++ {
			if n := len(lines[lineIndex]); n > max_ {
				max_ = n
			}
		}
		b := make([]byte, (max_+2)/3)
		for j := range b {
			b[j] = recognizeDigit(lines, j*3)
		}
		r[index] = string(b)
		lines = lines[4:]
	}
	return r
}

File: solutions/javascript/ocr-numbers/1/ocr-numbers.js
/**
 * OCR Patterns for each digit (0-9), represented as arrays of strings.
 */
const PATTERNS = {
  0: [' _ ', '| |', '|_|', '   '],
  1: ['   ', '  |', '  |', '   '],
  2: [' _ ', ' _|', '|_ ', '   '],
  3: [' _ ', ' _|', ' _|', '   '],
  4: ['   ', '|_|', '  |', '   '],
  5: [' _ ', '|_ ', ' _|', '   '],
  6: [' _ ', '|_ ', '|_|', '   '],
  7: [' _ ', '  |', '  |', '   '],
  8: [' _ ', '|_|', '|_|', '   '],
  9: [' _ ', '|_|', ' _|', '   '],
};

/**
 * Splits the text into rows with each row containing 4 lines.
 * @param {string} text - The input text representing OCR digits.
 * @returns {string[]} - An array of rows.
 */
const splitIntoRows = (text) => {
  const rows = [];
  const lines = text.split('\n');
  for (let rowNumber = 0; rowNumber < lines.length; rowNumber += 4) {
    let row = '';
    for (let rowLine = 0; rowLine < 4; rowLine += 1) {
      row += `${lines[rowNumber + rowLine]}\n`;
    }
    rows.push(row.slice(0, -1));
  }
  return rows;
};

/**
 * Splits a row into individual digits.
 * @param {string} row - A row of OCR digits.
 * @returns {string[]} - An array of digit strings.
 */
const splitIntoDigits = (row) => {
  const digits = [];
  const rows = row.split('\n');
  for (let digitNumber = 0; digitNumber < rows[0].length; digitNumber += 3) {
    let digit = '';
    for (let rowNumber = 0; rowNumber < rows.length; rowNumber += 1) {
      digit += rows[rowNumber].substring(digitNumber, digitNumber + 3);
    }
    digits.push(digit);
  }
  return digits;
};

/**
 * Converts a digit string to its corresponding number.
 * @param {string} text - The digit string to convert.
 * @returns {number|string} - The corresponding number or '?' if invalid.
 */
const getDigit = (text) => {
  const digit = Object.values(PATTERNS)
    .map((x) => x.join(''))
    .indexOf(text);
  return digit === -1 ? '?' : digit;
};

/**
 * Converts a row of OCR digits to a string of numbers.
 * @param {string} row - A row of OCR digits.
 * @returns {string} - A string of numbers.
 */
const valuesInRow = (row) => splitIntoDigits(row).map(getDigit).join('');

/**
 * Converts OCR text to a string of numbers, with rows separated by commas.
 * @param {string} text - The input text representing OCR digits.
 * @returns {string} - The converted string of numbers.
 */
export const convert = (text) => splitIntoRows(text).map(valuesInRow).join(',');
